### DISCUSSION TITLE
layer_num variable.

### DISCUSSION THREAD

**Discussion by JCBond** [OrcaSlicer/OrcaSlicer#4802] [Q&A]
According to the description of the "layer_num," variable the first Layer should be layer number 1
![image](https://github.com/SoftFever/OrcaSlicer/assets/48573000/c59a8be0-8647-4a10-adef-65c308bc7809)

But when I use this variable in the "Layer Change G-code," it seems that the first Layer is number 0. Is this a bug?

**Comment by TexZeTech** [comment]
> 
> But when I use this variable in the "Layer Change G-code," it seems that the first Layer is number 0. Is this a bug?

In short no.

This is why I use in my Layer Change G-code

`;AFTER_LAYER_CHANGE
;[layer_z]
SET_PRINT_STATS_INFO CURRENT_LAYER={layer_num + 1} 
M117 Layer {layer_num + 1}/[total_layer_count]`

Please edit to your use case.

**Reply by JCBond** [reply]
Thanks for the G-code, but the description needs to be updated to say that the first layer is layer 0. At the moment, it clearly says that the first layer should be layer 1.

**Reply by TexZeTech** [reply]
<img src="https://github.com/SoftFever/OrcaSlicer/assets/7254540/7fd28d93-d4de-4439-b5f1-0a08df930f20" width=80% height=80%/>

It's not a definition there its saying what the output can/will say. 

If it said "The first layer starts at 20" then I would agree.

**Reply by JCBond** [reply]
I understand what, i.e., means.
If you read the statement in the definition, it clearly says: "One-based," and then it explains what One-based... means: "in other words," the first layer is number one.

**Reply by TexZeTech** [reply]
You are correct and I apologize.

**Reply by Klober81** [reply]
Unfortunately, it's actually worse - depending on which custom G-code section you're using layer_num in it actually starts at -1. Put ";Layer#=[layer_num] - Z=[layer_z]" into the "Change filament G-code" section and check the sliced G-code - the result will be ";Layer#=-1 - Z=0". I just found this out while trying to troubleshoot an If statement where I'm using layer_num.
![image](https://github.com/SoftFever/OrcaSlicer/assets/20543393/5b2e7cce-f4d6-462d-8772-21282d7a2243)

**Reply by TexZeTech** [reply]
So on my klipperised printer I use the following code in the after layer change section.
```
SET_PRINT_STATS_INFO CURRENT_LAYER={layer_num + 1} 
M117 Layer {layer_num + 1}/[total_layer_count]
```

and that works so I'm curious, what firmware are you using and what is the macro you are using @Klober81 ?

**Reply by Klober81** [reply]
I'm using Klipper and I need to use layer_num in an If statement to adjust the Z offset based on the bed type for my Vivedino Marathon. Having the layer_num be different based on which custom G-code section it's being used in is unintuitive, and having it not actually be "One-based" as stated in the placeholder description is confusing until debug code is used to determine the "actual" layer number according to the slicer. This is particularly problematic when compounded by other issues I've run into like calculations and variable declarations still being calculated and declared even though they're commented out and e_retracted not seeming to work (I'll be adding issues for these today). I'm guessing these issues are carry-overs from Bambu Studio since it seems Bambu Lab did a somewhat hasty conversion of Prusa Slicer for their own needs.

**Reply by Klober81** [reply]
Spent all day on a workaround for this issue so I guess I'll be working on adding the other 2 issues tomorrow. 😂
